tidy3d/plugins/smatrix/ports/coaxial_lumped.py

```diff
@@ -0,0 +1,381 @@
+
+
+class CoaxialLumpedPort(AbstractLumpedPort):
+    """Class representing a single coaxial lumped port
+
+    Example
+    -------
+    >>> port1 = CoaxialLumpedPort(center=(0, 0, 0),
+    ...             outer_diameter=4,
+    ...             inner_diameter=1,
+    ...             normal_axis=2,
+    ...             direction="+",
+    ...             name="coax_port_1",
+    ...             impedance=50
+    ...         )
+    """
+
+    center: Coordinate = pd.Field(
+        (0.0, 0.0, 0.0),
+        title="Center",
+        description="Center of object in x, y, and z.",
+        units=MICROMETER,
+    )
+
+    outer_diameter: pd.PositiveFloat = pd.Field(
+        ...,
+        title="Outer Diameter",
+        description="Diameter of the outer coaxial circle.",
+        units=MICROMETER,
+    )
+
+    inner_diameter: pd.PositiveFloat = pd.Field(
+        ...,
+        title="Inner Diameter",
+        description="Diameter of the inner coaxial circle.",
+        units=MICROMETER,
+    )
+
+    normal_axis: Axis = pd.Field(
+        ...,
+        title="Normal Axis",
+        description="Specifies the axis which is normal to the concentric circles.",
+    )
+
+    direction: Direction = pd.Field(
+        ...,
+        title="Direction",
+        description="Specifies which side of the port the inner conductor is extending from. "
+        "This is needed in order to position the path integral, which is used for computing "
+        "conduction current using Ampère's circuital law.",
+    )
+
+    @cached_property
+    def injection_axis(self):
+        """Injection axis of the port."""
+        return self.normal_axis
+
+    @pd.validator("center", always=True)
+    def _center_not_inf(cls, val):
+        """Make sure center is not infinity."""
+        if any(np.isinf(v) for v in val):
+            raise ValidationError("Center can not contain td.inf terms.")
+        return val
+
+    @pd.validator("inner_diameter", always=True)
+    @skip_if_fields_missing(["outer_diameter"])
+    def _ensure_inner_diameter_is_smaller(cls, val, values):
+        """Ensures that the inner diameter is smaller than the outer diameter, so that the final
+        shape is an annulus."""
+        outer_diameter = values.get("outer_diameter")
+        if val >= outer_diameter:
+            raise ValidationError(
+                f"The inner diameter {val} of a coaxial lumped element must be less than the "
+                "outer diameter {outer_diameter}."
+            )
+        return val
+
+    def to_source(
+        self, source_time: GaussianPulse, snap_center: float, grid: Grid
+    ) -> CustomCurrentSource:
+        """Create a current source from the lumped port."""
+        # Discretized source amps are manually zeroed out later if they
+        # fall on Yee grid locations outside the analytical source region.
+
+        # Get transverse axes
+        trans_axes = self.remaining_axes
+
+        coord1 = "xyz"[trans_axes[0]]
```

Comment:
possible applications of `pop_axis` here and below, eg.
```py
coord1, (coord2, coord3) = pop_axis('xyz', axis=self.normal_axis)
```

Reply:
The only thing that prevents me from using pop_axis here is I find it easier to always remain in a 2D coordinate system that is equivalent to "xy". If "y" is removed I prefer "zx" as the order of coordinates so that I don't need to introduce any minus signs when performing contour integrals. That is why I have that logic in `self.remaining_axes`.

But I can add a method that will act like `pop_axis` but returns the order that I find convenient.

Reply:
I see. Hm, if you find the other method convenient and can re-use it in other places, feel free. otherwise feel free to ignore.